OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,457 @@
+
+		const uint LinkPrev          = 0x1;
+		const uint LinkNext          = 0x2;
+		const uint MaxFileGroupCount = 71;
+#region Nested Structs
+		struct FileGroup
+		{
+			public readonly string Name;
+			public readonly uint FirstFile;
+			public readonly uint LastFile;
+			public FileGroup(BinaryReader reader, long offset)
+			{
+				var nameOffset = reader.ReadUInt32();
+				/*   unknown  */ reader.ReadBytes(0x12);
+				FirstFile      = reader.ReadUInt32();
+				LastFile       = reader.ReadUInt32();
+
+				reader.BaseStream.Seek(offset + (long)nameOffset, SeekOrigin.Begin);
+				Name = reader.BaseStream.ReadASCIIZ();
+			}
+		}
+
+		struct VolumeHeader
+		{
+			public uint DataOffset;
+			public uint DataOffsetHigh;
+			public uint FirstFileIndex;
+			public uint LastFileIndex;
+
+			public uint FirstFileOffset;
+			public uint FirstFileOffsetHigh;
+			public uint FirstFileSizeExpanded;
+			public uint FirstFileSizeExpandedHigh;
+
+			public uint FirstFileSizeCompressed;
+			public uint FirstFileSizeCompressedHigh;
+			public uint LastFileOffset;
+			public uint LastFileOffsetHigh;
+
+			public uint LastFileSizeExpanded;
+			public uint LastFileSizeExpandedHigh;
+			public uint LastFileSizeCompressed;
+			public uint LastFileSizeCompressedHigh;
+
+			public VolumeHeader(BinaryReader reader)
+			{
+				DataOffset      = reader.ReadUInt32();
+				DataOffsetHigh  = reader.ReadUInt32();
+
+				FirstFileIndex      = reader.ReadUInt32();
+				LastFileIndex       = reader.ReadUInt32();
+				FirstFileOffset     = reader.ReadUInt32();
+				FirstFileOffsetHigh = reader.ReadUInt32();
+
+				FirstFileSizeExpanded       = reader.ReadUInt32();
+				FirstFileSizeExpandedHigh   = reader.ReadUInt32();
+				FirstFileSizeCompressed     = reader.ReadUInt32();
+				FirstFileSizeCompressedHigh = reader.ReadUInt32();
+
+				LastFileOffset           = reader.ReadUInt32();
+				LastFileOffsetHigh       = reader.ReadUInt32();
+				LastFileSizeExpanded     = reader.ReadUInt32();
+				LastFileSizeExpandedHigh = reader.ReadUInt32();
+
+				LastFileSizeCompressed     = reader.ReadUInt32();
+				LastFileSizeCompressedHigh = reader.ReadUInt32();
+			}
+		}
+
+		struct CommonHeader
+		{
+			public readonly uint Version;
+			public readonly uint VolumeInfo;
+			public readonly long CabDescriptorOffset;
+			public readonly uint CabDescriptorSize;
+
+			public CommonHeader(BinaryReader reader)
+			{
+				Version             = reader.ReadUInt32();
+				VolumeInfo          = reader.ReadUInt32();
+				CabDescriptorOffset = (long)reader.ReadUInt32();
+				CabDescriptorSize   = reader.ReadUInt32();
+			}
+		}
+
+		struct CabDescriptor
+		{
+			public readonly long FileTableOffset;
+			public readonly uint FileTableSize;
+			public readonly uint FileTableSize2;
+			public readonly uint DirectoryCount;
+
+			public readonly uint FileCount;
+			public readonly long FileTableOffset2;
+
+			public CabDescriptor(BinaryReader reader, CommonHeader commonHeader)
```

Comment:
Likewise `BinaryReader` &rarr; `Stream` here and everywhere else in this parser.